### PR TITLE
F7832 fix logins storage exception

### PR DESCRIPTION
**Pull request by Baron-Severin**
---

### Pull Request checklist

- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [X] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/AsyncLoginsStorageAdapterTest.kt [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                unlock(key().await())
+                list().await()
+                lock()
+            }
+        }
+
+        runBlocking {
+            withTimeoutOrNull(100) {
+                for (e in exceptionChannel) {
+                    throw e
```
Comment (Baron-Severin) [inline components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/AsyncLoginsStorageAdapterTest.kt]: This was the best way I could come up with to reproduce this in a test.  I'm open to suggestions.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt [hunk 2]
```diff
@@ -422,6 +442,26 @@ open class AsyncLoginsStorageAdapter<T : LoginsStorage>(private val wrapped: T)
...
+        try {
+            ensureUnlocked(key().await()).await()
+
+            try {
+                return block(this)
+            } finally {
+                ensureLocked().await()
+            }
+        } finally {
+            mutex.unlock(wrapped)
```
Comment (Baron-Severin) [inline components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt]: Also open to alternatives here.  This is very ugly.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/LoginDialogFragment.kt [hunk 3]
```diff
@@ -197,7 +199,9 @@ internal class LoginDialogFragment : PromptDialogFragment() {
             username = username,
             password = password
         )
-        stateUpdate?.cancel()
+        // The below code is not thread safe, so we ensure that we block here until sure that the
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/LoginDialogFragment.kt]: nit: "unit we are sure"

**Comment by Baron-Severin** [comment]
bors r=grigoryk

**Pull request merged**